modular_skyrat/modules/modular_implants/code/soulcatcher/handheld_soulcatcher.dm

```diff
@@ -47,68 +56,112 @@
-	window_flash(target_mob.client)
+	return TRUE
 
-	if((tgui_alert(target_mob, "Do you wish to enter [target_room]? This will remove you from your body until you leave.", name, list("Yes", "No"), 30 SECONDS, FALSE) != "Yes") || (tgui_alert(target_mob, "Are you sure about this?", name, list("Yes", "No"), 30 SECONDS, FALSE) != "Yes"))
-		COOLDOWN_START(src, rsd_scan_cooldown, RSD_ATTEMPT_COOLDOWN)
-		to_chat(user, span_warning("[target_mob] doesn't seem to want to enter."))
+/**
+ * Invites target_mob into target_room by giving them a tgui_alert.
+ * target_mob cannot be invited more than once while a alert from a given soulcatcher is open, enforced by
+ * src.confirming_entry += real_target and a check surrounding that.
+ *
+ * Args:
+ * datum/soulcatcher/target_room: The room we are inviting target_mob to.
+ * mob/living/user (Nullable): The individual who invited the mob.
+ * mob/living/target_mob: The mob that was clicked on/invited into the room.
+ *
+ * Returns:
+ * False if the invitation wasn't delivered, the result of the tgui_alert otherwise.
+ */
+/obj/item/handheld_soulcatcher/proc/invite_soul(datum/soulcatcher_room/target_room, mob/living/user, mob/living/target_mob)
+	if (target_mob in confirming_entry)
+		if (user)
+			to_chat(user, span_warning("You've already invited this person to a room, wait for them to respond!"))
 		return FALSE
 
-	if(!target_mob.mind)
-		return FALSE
+	var/message = "Do you want to transfer your soul into [target_room]?"
+	if (user)
+		message = "[user] wants to transfer you to [target_room] inside of a soulcatcher, do you accept?"
+	var/target_dead = (target_mob.stat == DEAD)
+	var/mob/dead/observer/ghost = target_mob.get_ghost(TRUE, TRUE)
+	var/mob/real_target = target_mob
 
-	target_room.add_soul(target_mob.mind, TRUE)
-	playsound(src, 'modular_skyrat/modules/modular_implants/sounds/default_good.ogg', 50, FALSE, ignore_walls = FALSE)
-	visible_message(span_notice("[src] beeps: [target_mob]'s mind transfer is now complete."))
+	if(ghost || target_dead) //We can temporarily store souls of dead mobs.
+		target_mob.ghostize(TRUE) //Incase they are staying in the body.
+		ghost = target_mob.get_ghost(TRUE, TRUE) // ghostized, new ghost could be here
+		if(!ghost)
+			if (user)
+				to_chat(user, span_warning("You are unable to get the soul of [target_mob]!"))
+			return FALSE
+		else
+			real_target = ghost
+	else
+		message += " This will remove you from your body until you leave."
 
-	if(!target_mob.GetComponent(/datum/component/previous_body))
-		return FALSE
+	confirming_entry += real_target
+	RegisterSignal(real_target, COMSIG_QDELETING, PROC_REF(handle_confirming_soul_del))
+	SEND_SOUND(real_target, 'sound/misc/notice2.ogg')
+	window_flash(real_target.client)
+	var/invitation_results = (((tgui_alert(real_target, message, name, list("Yes", "No"), 30 SECONDS) == "Yes") && (tgui_alert(real_target, "Are you sure about this?", name, list("Yes", "No"), 30 SECONDS) == "Yes")))
```

Comment:
This is really cursed and ugly, please don't do this. Make this into two proper separate ifs. Just call `handle_confirming_soul_del()` manually if you're going to return false. It's going to be far more readable than this.

Reply:
worth noting this wasnt my code, this was pasted from the OG soulcatcher code

and yes i agree its ugly i shouldve replaced it with something better

Reply:
(i will)